src/solver/linear/preconditioners/l1_gauss_seidel.jl

```diff
@@ -0,0 +1,310 @@
+
+"""
+struct L1GSPrecBuilder
+    backend::Backend
+    function L1GSPrecBuilder(backend::Backend)
+        if functional(backend)
+            return new(backend)
+        else
+            error(" $backend is not functional, please check your backend.")
+        end
+    end
+end
+
+(builder::L1GSPrecBuilder)(A::AbstractMatrix,partsize::Ti,nparts::Ti) where {Ti} = 
+    build_l1prec(builder, A,partsize, nparts)
+
+  
+struct DiagonalIterator{MatrixFormat,MatrixSymmetry <: AbstractMatrixSymmetry,MatrixType,VectorType,Ti} 
+    A::MatrixType
+    B::VectorType # Diagonal values
+    D::VectorType # Off-diagonal absolute sum
+    partsize::Ti
+    nparts::Ti
+    local_idx::Ti # local index never changes
+    initial_partition_idx::Ti # why initial ? because we are using stride to loop over the diagonals. 
+    initial_global_idx::Ti 
+    function DiagonalIterator(::Type{symT}, A::MatrixType,B::VectorType,D::VectorType,partsize::Ti,nparts::Ti,local_idx::Ti,initial_partition_idx::Ti) where {symT<:AbstractMatrixSymmetry,MatrixType,VectorType,Ti}
+        format_type = sparsemat_format_type(A) 
+        initial_global_idx = (initial_partition_idx - convert(Ti,1)) * partsize + local_idx
+        new{format_type,symT,MatrixType,VectorType,Ti}(A,B,D,partsize,nparts,local_idx,initial_partition_idx,initial_global_idx)
+    end
+end
+
+struct DiagonalCache{Ti,Tv}
+    k::Ti # partition index
+    idx::Ti # diagonal index
+    b::Tv # partition diagonal value
+    d::Tv # off-partition absolute sum
+end
+
+## Preconditioner builder ##
+function build_l1prec(builder::L1GSPrecBuilder, A::MatrixType,partsize::Ti, nparts::Ti) where {Ti<:Integer,MatrixType}
+    @unpack backend = builder
+    partsize == 0 && error("partsize must be greater than 0")
+    nparts == 0 && error("nparts must be greater than 0")
+    _build_l1prec(backend, A, partsize, nparts)
+end
+
+function _build_l1prec(backend::Backend, _A::MatrixType,partsize::Ti, nparts::Ti) where {Ti<:Integer,MatrixType}
+    # No assumptions on A, i.e. A here might be in either backend compatible format or not. 
+    # So we have to convert it to backend compatible format, if it is not already.
+    partitioning = BlockPartitioning(partsize, nparts, backend)
+    A = get_data(_A) # for symmetric case
+    B,D = _precompute_blocks(A, partitioning)
+    L1GSPreconditioner(partitioning, B,D)
+end
+
+function LinearSolve.ldiv!(y::VectorType, P::L1GSPreconditioner{BlockPartitioning{Ti,Backend}}, x::VectorType) where {VectorType <: AbstractVector, Ti<:Integer, Backend}
+    # x: residual
+    # y: preconditioned residual
+    y .= x #works either way, whether x is GpuVectorType (e.g. CuArray) or Vector
+    @unpack partitioning,B,D = P
+    @unpack backend = partitioning
+    # The following code is required because there is no assumption on the compatibality of x with the backend.
+    _y = adapt(backend, y)
+    _y .= _y ./ (B + D)
```

Comment:
Does

```suggestion
    @.. _y = _y / (B + D)
```

Change anything?